Discord voice can play TTS after session leave

## Bug type

Behavior bug (incorrect output/state without crash)

## Beta release blocker

No

## Summary

Discord voice TTS that finishes after its voice session is left can still start playback on the stopped session, and file-backed playback artifacts are retained after use.

## Steps to reproduce

1. Join a Discord voice session in `stt-tts` mode using a controlled Discord runtime fixture.
2. Start a voice turn whose TTS stream remains unresolved.
3. Leave the voice session while synthesis is still pending.
4. Let synthesis complete and allow the queued playback promise to settle.
5. Observe that `player.play()` is called once after leave.

Focused regression command:

```shell
pnpm test extensions/discord/src/voice/voice-runtime.e2e.test.ts
```

Before the repair, the focused suite reported 1 failed and 28 passed tests; the failure showed one post-leave `player.play()` call.

## Expected behavior

Leaving or stopping a voice session is terminal for that session's pending TTS completion and queued playback. Late work is discarded, provider streams are released, and file-backed playback artifacts are removed after success, discard, or error.

## Actual behavior

The session owner stops the current player, but synthesis and the playback queue can settle later and call `player.play()` again. The file-backed branch also leaves its generated media-store file behind after ephemeral voice playback.

## OpenClaw version

`2026.8.1`, reproduced from the current implementation carried through `6e5bf3ec55cb2c24ad74442d08ca68f51485ad23`.

## Operating system

Ubuntu 24.04 Blacksmith Testbox for the deterministic reproduction; source contract also inspected on macOS.

## Install method

Source checkout with the repository test harness.

## Model

N/A — controlled TTS fixture; no external model.

## Provider / routing chain

Controlled Discord voice runtime → delayed mock speech stream → real playback ownership path.

## Additional provider/model setup details

No credentials, real Discord send, or external speech provider are required. This isolates session lifecycle and playback ownership from provider behavior.

## Logs

```shell
FAIL  extensions/discord/src/voice/voice-runtime.e2e.test.ts
AssertionError: expected player.play not to be called, but it was called 1 time
Tests: 1 failed, 28 passed
```

## Screenshots, recordings, and evidence

The direct dependency contract confirms why `player.stop()` is not sufficient:

- `@discordjs/voice` 0.19.2 documents that `AudioPlayer` instances are reusable and unaffected by voice connections becoming unavailable: https://github.com/discordjs/discord.js/blob/2a067216c410e402f3ada7cd9eb29f566636b0ad/packages/voice/src/audio/AudioPlayer.ts#L204-L213
- `play()` accepts a new resource and transitions the player back into playback: https://github.com/discordjs/discord.js/blob/2a067216c410e402f3ada7cd9eb29f566636b0ad/packages/voice/src/audio/AudioPlayer.ts#L365-L452
- `stop()` ends only the current resource: https://github.com/discordjs/discord.js/blob/2a067216c410e402f3ada7cd9eb29f566636b0ad/packages/voice/src/audio/AudioPlayer.ts#L485-L500

Related but distinct: #116201 tracks provider-native realtime voice retention; this report concerns the synchronous Discord STT→TTS playback queue.

## Impact and severity

- Affected: Discord voice sessions using `stt-tts` mode.
- Severity: High; audio can start after the operator has left or replaced the session.
- Frequency: Deterministic when leave wins the synthesis/queue race.
- Consequence: stale audible output and retained generated media artifacts after the owning session is terminal.

## Additional information

The queue split landed in #122479, but the existing playback body was carried forward without revalidating the session after awaited synthesis or at queued execution time. No config, auth, protocol, SDK, delivery/replay, or provider contract change is needed.

--- a/extensions/discord/src/voice/segment.ts
+++ b/extensions/discord/src/voice/segment.ts
@@ -1,7 +1,7 @@
 // Discord plugin module implements segment behavior.
-import path from "node:path";
 import { Readable } from "node:stream";
 import type { DiscordAccountConfig, OpenClawConfig } from "openclaw/plugin-sdk/config-contracts";
+import { unlinkIfExists } from "openclaw/plugin-sdk/media-runtime";
 import type { RuntimeEnv } from "openclaw/plugin-sdk/runtime-env";
 import { createSubsystemLogger } from "openclaw/plugin-sdk/runtime-env";
 import { formatErrorMessage } from "openclaw/plugin-sdk/ssrf-runtime";
@@ -171,32 +171,35 @@ export async function processDiscordVoiceSegment(params: {
     `tts ok (${voiceReplyAudio.speakText.length} chars): guild ${entry.guildId} channel ${entry.channelId}`,
   );
 
+  const releaseAudio =
+    voiceReplyAudio.mode === "stream"
+      ? voiceReplyAudio.release
+      : () => unlinkIfExists(voiceReplyAudio.audioPath);
+  // Synthesis can settle after leave; release before the playback queue gets ownership.
+  if (entry.sessionLifecycle.status === "stopped") {
+    await releaseAudio?.();
+    return;
+  }
   params.enqueuePlayback(entry, async () => {
     const voiceSdk = loadDiscordVoiceSdk();
-    const releaseAudioStream =
-      voiceReplyAudio.mode === "stream" ? voiceReplyAudio.release : undefined;
     try {
-      if (voiceReplyAudio.mode === "stream") {
-        logVoiceVerbose(`playback start: guild ${entry.guildId} channel ${entry.channelId} stream`);
-        const nodeStream = Readable.fromWeb(
-          voiceReplyAudio.audioStream as import("node:stream/web").ReadableStream<Uint8Array>,
-        );
-        const resource = voiceSdk.createAudioResource(createDiscordOpusPlaybackStream(nodeStream), {
-          inputType: voiceSdk.StreamType.Opus,
-        });
-        entry.player.play(resource);
-      } else {
-        logVoiceVerbose(
-          `playback start: guild ${entry.guildId} channel ${entry.channelId} file ${path.basename(voiceReplyAudio.audioPath)}`,
-        );
-        const resource = voiceSdk.createAudioResource(
-          createDiscordOpusPlaybackStream(voiceReplyAudio.audioPath),
-          {
-            inputType: voiceSdk.StreamType.Opus,
-          },
-        );
-        entry.player.play(resource);
+      // Queued playback can outlive its session; a stopped player is reusable by the SDK.
+      if (entry.sessionLifecycle.status === "stopped") {
+        return;
       }
+      const input =
+        voiceReplyAudio.mode === "stream"
+          ? Readable.fromWeb(
+              voiceReplyAudio.audioStream as import("node:stream/web").ReadableStream<Uint8Array>,
+            )
+          : voiceReplyAudio.audioPath;
+      logVoiceVerbose(
+        `playback start: guild ${entry.guildId} channel ${entry.channelId} ${voiceReplyAudio.mode}`,
+      );
+      const resource = voiceSdk.createAudioResource(createDiscordOpusPlaybackStream(input), {
+        inputType: voiceSdk.StreamType.Opus,
+      });
+      entry.player.play(resource);
       await voiceSdk
         .entersState(entry.player, voiceSdk.AudioPlayerStatus.Playing, PLAYBACK_READY_TIMEOUT_MS)
         .catch(() => undefined);
@@ -205,7 +208,7 @@ export async function processDiscordVoiceSegment(params: {
         .catch(() => undefined);
       logVoiceVerbose(`playback done: guild ${entry.guildId} channel ${entry.channelId}`);
     } finally {
-      await releaseAudioStream?.();
+      await releaseAudio?.();
     }
   });
 }

--- a/extensions/discord/src/voice/voice-runtime.e2e.test.ts
+++ b/extensions/discord/src/voice/voice-runtime.e2e.test.ts
@@ -1,3 +1,6 @@
+import fs from "node:fs/promises";
+import os from "node:os";
+import path from "node:path";
 import type { MockCallSource } from "./manager.e2e.test-support.js";
 import { defineDiscordVoiceTests } from "./voice-test-harness.test-support.js";
 
@@ -220,6 +223,7 @@ defineDiscordVoiceTests(
           route: { sessionKey: "discord:g1:1001", agentId: "agent-1" },
           connection: createConnectionMock(),
           player: createAudioPlayerMock(),
+          sessionLifecycle: { status: "active" },
           playbackQueue: Promise.resolve(),
           processingQueue: Promise.resolve(),
           capture: createVoiceCaptureState(),
@@ -282,6 +286,10 @@ defineDiscordVoiceTests(
     });
 
     it("runs voice replies under Discord voice output policy", async () => {
+      const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "openclaw-discord-voice-"));
+      const audioPath = path.join(tempDir, "reply.mp3");
+      await fs.writeFile(audioPath, "voice");
+      textToSpeechMock.mockResolvedValueOnce({ success: true, audioPath });
       agentCommandMock.mockResolvedValueOnce({
         payloads: [{ text: "hello back" }],
       } as never);
@@ -292,7 +300,18 @@ defineDiscordVoiceTests(
         client,
         {},
       );
-      await processVoiceSegment(manager, "u-guest");
+      try {
+        await processVoiceSegment(manager, "u-guest");
+        await vi.waitFor(async () => {
+          const exists = await fs.access(audioPath).then(
+            () => true,
+            () => false,
+          );
+          expect(exists).toBe(false);
+        });
+      } finally {
+        await fs.rm(tempDir, { recursive: true, force: true });
+      }
 
       const commandArgs = lastAgentCommandArgs() as
         | { message?: string; messageChannel?: string; messageProvider?: string }
@@ -366,6 +385,51 @@ defineDiscordVoiceTests(
         throw new Error("expected Discord audio resource input");
       }
       await vi.waitFor(() => expect(release).toHaveBeenCalledTimes(1));
+    });
+
+    it("releases late TTS without playback after the voice session leaves", async () => {
+      const connection = createConnectionMock();
+      joinVoiceChannelMock.mockReturnValueOnce(connection);
+      decodeOpusStreamMock.mockResolvedValueOnce(Buffer.alloc(96_000));
+      agentCommandMock.mockResolvedValueOnce({ payloads: [{ text: "late voice reply" }] });
+      const release = vi.fn(async () => undefined);
+      let resolveStream!: (value: unknown) => void;
+      textToSpeechStreamMock.mockReturnValueOnce(
+        new Promise((resolve) => {
+          resolveStream = resolve;
+        }),
+      );
+      connection.receiver.subscribe.mockReturnValueOnce({
+        on: vi.fn(),
+        off: vi.fn(),
+        destroy: vi.fn(),
+        destroyed: false,
+        async *[Symbol.asyncIterator]() {},
+      });
+      const manager = createManager(
+        makeVoiceConfig({}, { groupPolicy: "open", allowFrom: ["discord:u-speaker"] }),
+      );
+      await manager.join({ guildId: "g1", channelId: "1001" });
+      const entry = getSessionEntry(manager);
+
+      const speaking = handleSpeakingStart(manager, entry, "u-speaker");
+      await vi.waitFor(() => expect(textToSpeechStreamMock).toHaveBeenCalledOnce());
+      await manager.leave({ guildId: "g1" });
+      resolveStream({
+        success: true,
+        audioStream: new ReadableStream<Uint8Array>({
+          start(controller) {
+            controller.close();
+          },
+        }),
+        release,
+      });
+      await speaking;
+      await entry.processingQueue;
+      await entry.playbackQueue;
+
+      expect(entry.player.play).not.toHaveBeenCalled();
+      expect(release).toHaveBeenCalledOnce();
     });
 
     it("passes per-channel system prompt context to voice agent runs", async () => {

--- a/extensions/discord/src/voice/voice-test-harness.test-support.ts
+++ b/extensions/discord/src/voice/voice-test-harness.test-support.ts
@@ -550,6 +550,7 @@ function buildVoiceTestHarness() {
         route: { sessionKey: "discord:g1:1001", agentId: "agent-1" },
         connection: createConnectionMock(),
         player: createAudioPlayerMock(),
+        sessionLifecycle: { status: "active" },
         playbackQueue: Promise.resolve(),
         processingQueue: Promise.resolve(),
         capture: createVoiceCaptureState(),

